### PR TITLE
Editing calendar

### PR DESCRIPTION
**Pull request by chioma-ibe**
**PR 07.21.2025**

**What's in this PR**

This PR introduces study session preference management and a recommendation algorithm that finds compatible study times between users based on their availability and preferred study patterns. It also adds a reusable spinner component for improved UI during loading states.

**Study Preferences & Matching Algorithm**

Created a new backend module called studyPreferences.js to handle storing and retrieving users’ study preferences and finding optimal session times between users.

- createOrUpdateStudyPreferences stores preferred days, time ranges, back-to-back session preference, max sessions per week, and session duration in the database.
- getStudyPreferences Fetches saved preferences for a user.
- getCompatibleStudyTimes is the matching algorithm that returns up to 3 optimal time slots based on shared preferred days, overlapping time ranges, session duration, and back-to-back preference.

In my algorithm, if no overlapping time ranges are found it suggest sessions on common days using only one user’s time ranges and if users have no common days, suggest based on one user’s days and the other’s times and as a last resort, suggests based only on one user’s full availability

**Reusable Spinner Component**

- Created a styled Spinner component with basic loading animation
- Integrated the spinner into loading states across key pages to provide better feedback

**Testing done** 

- Added a script with multiple test users to validate the algorithm against different scenarios including full overlap, partial overlap, and no overlap.
- Verified that fallback logic is triggered correctly and sessions are never scheduled in the past.
- Validated that session duration is respected regardless of how wide the availability window is.
- Confirmed that back-to-back sessions are only created if both users prefer them.
- Tested preference creation and update across multiple users
- Validated that perfect matches are prioritized

(Test script is included in the PR) 

Here are the results that were logged from running the script

TEST CASE 1: Users with partial weekday overlap
User 1 (Morning Weekday) + User 2 (Afternoon Weekday)
Expected: Perfect matches on Monday and Friday afternoons (1pm-3pm)
Compatible times found: 2
[
  {
    "start": "2025-07-28T20:00:00.000Z",
    "end": "2025-07-28T21:00:00.000Z",
    "day": "MONDAY",
    "perfectMatch": true
  },
  {
    "start": "2025-07-25T20:00:00.000Z",
    "end": "2025-07-25T21:00:00.000Z",
    "day": "FRIDAY",
    "perfectMatch": true
  }
]

TEST CASE 2: Users with no common days
User 1 (Morning Weekday) + User 4 (Weekend Only)
Expected: No perfect matches, only suggestions
Compatible times found: 3
[
  {
    "start": "2025-07-23T16:00:00.000Z",
    "end": "2025-07-23T17:00:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": false
  },
  {
    "start": "2025-07-23T20:00:00.000Z",
    "end": "2025-07-23T21:00:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": false
  },
  {
    "start": "2025-07-25T16:00:00.000Z",
    "end": "2025-07-25T17:00:00.000Z",
    "day": "FRIDAY",
    "perfectMatch": false
  }
]

TEST CASE 3: Users with very limited overlap
User 1 (Morning Weekday) + User 5 (Limited Availability)
Expected: One perfect match on Wednesday 2pm-3pm
Compatible times found: 1
[
  {
    "start": "2025-07-23T21:00:00.000Z",
    "end": "2025-07-23T21:30:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": true
  }
]

TEST CASE 4: Users with common days but no time overlap
User 1 (Morning Weekday) + User 6 (Night Owl)
Expected: No perfect matches despite common days
Compatible times found: 3
[
  {
    "start": "2025-07-23T16:00:00.000Z",
    "end": "2025-07-23T17:00:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": false,
    "reason": "Common day but using your preferred time"
  },
  {
    "start": "2025-07-23T20:00:00.000Z",
    "end": "2025-07-23T21:00:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": false,
    "reason": "Common day but using your preferred time"
  },
  {
    "start": "2025-07-25T16:00:00.000Z",
    "end": "2025-07-25T17:00:00.000Z",
    "day": "FRIDAY",
    "perfectMatch": false,
    "reason": "Common day but using your preferred time"
  }
]

TEST CASE 5: Limited availability + wide availability
User 5 (Limited Availability) + User 7 (Always Available)
Expected: Perfect matches on Wednesday 2pm-4pm
Compatible times found: 1
[
  {
    "start": "2025-07-23T21:00:00.000Z",
    "end": "2025-07-23T21:30:00.000Z",
    "day": "WEDNESDAY",
    "perfectMatch": true
  }
]

TEST CASE 6: Users with different session duration preferences
User 2 (90 min sessions) + User 3 (45 min sessions)
Expected: Compatible times with 45-minute duration
Compatible times found: 2
[
  {
    "start": "2025-07-22T19:00:00.000Z",
    "end": "2025-07-22T19:45:00.000Z",
    "day": "TUESDAY",
    "perfectMatch": true
  },
  {
    "start": "2025-07-24T19:00:00.000Z",
    "end": "2025-07-24T19:45:00.000Z",
    "day": "THURSDAY",
    "perfectMatch": true
  }
]

TEST CASE 7: Users who both prefer back-to-back sessions
User 3 (back-to-back) + User 7 (back-to-back)
Expected: Multiple consecutive time slots
Compatible times found: 3
[
  {
    "start": "2025-07-27T17:00:00.000Z",
    "end": "2025-07-27T17:45:00.000Z",
    "day": "SUNDAY",
    "perfectMatch": true
  },
  {
    "start": "2025-07-22T17:00:00.000Z",
    "end": "2025-07-22T17:45:00.000Z",
    "day": "TUESDAY",
    "perfectMatch": true
  },
  {
    "start": "2025-07-24T17:00:00.000Z",
    "end": "2025-07-24T17:45:00.000Z",
    "day": "THURSDAY",
    "perfectMatch": true
  }
]
These logs verify my algorithm works and returns the expected results. 

Confirmed that the spinner displays properly in loading states

https://www.loom.com/share/8698f1527b2d40e8a0ba84b82bd43a22?sid=6f60a90d-19b7-450e-b11d-13e617ae5638

[Vite + React - 21 July 2025 - Watch Video](https://www.loom.com/share/8698f1527b2d40e8a0ba84b82bd43a22)

**Blockers**

- Took some time to design the algorithm and debug fallback behavior so the algorithm wouldn’t suggest invalid sessions, and only display the sessions based on the expected result
- Setting up realistic test data for edge cases like partial overlaps or preference gaps helped improve confidence in results

**What is coming in future PR** 

- Iterating over tc 1: Adding dynamic weights and more factors that influence recommendations 
- Adding handshake functionality to application to confirm matches

### CODE REVIEW

File: backend/services/studyPreferencesService.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+        overlappingRanges.push({
+          start: overlapStart,
+          end: overlapEnd,
+          perfectMatch: true
+        });
+      }
+    }
+  }
+
```
Comment (hassainsaeed) [inline backend/services/studyPreferencesService.js:84]: This is a O(n^2) operation. I think we can make this more efficient

Are user1TimeRanges and user2TimeRanges already sorted? If so then you don't need to loop through every time range for user1 and time range for user2. You can start at the first time ranges for both users, check if theres overlap (if so, add it to overlappingRanges), and then increment the index on whichever time range ends first.

File: backend/services/studyPreferencesService.js [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        const availableMs = endTime - startTime;
+
+        if (availableMs >= durationMs) {
+          suggestedTimes.push({
+            start: startTime.toISOString(),
+            end: new Date(startTime.getTime() + durationMs).toISOString(),
+            day,
+            perfectMatch: false,
+          });
+        }
```
Comment (hassainsaeed) [inline backend/services/studyPreferencesService.js:319]: This block of code seems very similar to the block of code from L235 - L271. Is this intentional? If so, might be good to put this in a util/helper function to save repeated lines of code

File: frontend/src/components/spinner/Spinner.jsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import '../../styles/components/spinner/Spinner.css';
+
+const Spinner = () => {
+  return (
+    <div className="spinner-container">
+      <div className="spinner"></div>
+    </div>
+  );
+};
```
Comment (hassainsaeed) [inline frontend/src/components/spinner/Spinner.jsx:9]: Love this!

File: backend/scripts/testStudyPreferences.js [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    user2: testUser2.id,
+    user3: testUser3.id,
+    user4: testUser4.id,
+    user5: testUser5.id,
+    user6: testUser6.id,
+    user7: testUser7.id
+  };
+}
+
+async function testCompatibleTimes(userIds) {
```
Comment (hassainsaeed) [inline backend/scripts/testStudyPreferences.js:195]: Love that you have a test for the scheduling algorithm, that is amazing!
